### PR TITLE
multiraft: use fewer pointers

### PR DESCRIPTION
**Pull request by tamird**
Lots of pointers here that don't seem to be necessary. Not sure this fixes any issues, but came upon this while looking at #2639.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3117)

### CODE REVIEW

File: multiraft/multiraft.go [hunk 1]
```diff
@@ -644,10 +644,7 @@ func (s *state) start() {
 	})
 }
 
-func (s *state) removePending(g *group, prop *proposal, err error) {
-	if prop == nil {
-		return
-	}
+func (s *state) removePending(g *group, prop proposal, err error) {
```
Comment (bdarnell) [inline multiraft/multiraft.go]: This method mutates `prop`, so this one needs to stay a pointer, as does the `pending` map.

File: multiraft/multiraft.go [hunk 2]
```diff
@@ -159,10 +159,10 @@ func NewMultiRaft(nodeID roachpb.NodeID, storeID roachpb.StoreID, config *Config
 		Events: make(chan []interface{}),
 
 		// Input channels.
-		reqChan:         make(chan *RaftMessageRequest, reqBufferSize),
-		createGroupChan: make(chan *createGroupOp),
-		removeGroupChan: make(chan *removeGroupOp),
-		proposalChan:    make(chan *proposal),
+		reqChan:         make(chan RaftMessageRequest, reqBufferSize),
```
Comment (bdarnell) [inline multiraft/multiraft.go]: Buffered channels statically allocate their buffer, so changing a buffered channel from pointers to values has a larger memory cost. Also, using a shallow copy of protobufs can be misleading, so I would keep this one a pointer.

File: multiraft/multiraft.go (hunk 1 above)
Comment (tamird) [inline multiraft/multiraft.go]: Should that have made some tests fail?

Comment (bdarnell) [inline multiraft/multiraft.go]: Looks like it did: https://circleci.com/gh/cockroachdb/cockroach/9287

Comment (tamird) [inline multiraft/multiraft.go]: I don't think so, that's #2639

Comment (bdarnell) [inline multiraft/multiraft.go]: I'm not sure; the log messages are slightly different and I think "cannot remove unknown node 1" is a consequence of this change. But yes, a more direct and deterministic test of this functionality would be valuable.

File: multiraft/multiraft.go (hunk 2 above)
Comment (tamird) [inline multiraft/multiraft.go]: Done.

File: multiraft/multiraft.go (hunk 1 above)
Comment (tamird) [inline multiraft/multiraft.go]: Reworked this a bit.

**Comment by tamird** [comment]
This turned out less interesting than I thought :(

File: multiraft/multiraft.go [hunk 3]
```diff
@@ -644,10 +644,7 @@ func (s *state) start() {
 	})
 }
 
-func (s *state) removePending(g *group, prop *proposal, err error) {
-	if prop == nil {
-		return
-	}
+func (prop *proposal) removeFromPending(g *group, err error) {
```
Comment (bdarnell) [inline multiraft/multiraft.go]: Why this change?

Comment (tamird) [inline multiraft/multiraft.go]: two reasons: `s` was never used in this method, and having a pointer receiver makes it clearly (to me?) that it's being mutated

Comment (bdarnell) [inline multiraft/multiraft.go]: But `g` is also being mutated.

Comment (tamird) [inline multiraft/multiraft.go]: Yeah, you're right, it probably makes more sense to make that the receiver. Done.
